### PR TITLE
[DatePicker] Refactor tests to testing-library

### PR DESCRIPTION
**Pull request by dmtrKovalenko**

### CODE REVIEW

File: .mocharc.js [hunk 1]
```diff
@@ -1,5 +1,8 @@
 module.exports = {
   recursive: true,
   reporter: 'dot',
-  require: [require.resolve('@babel/register'), require.resolve('./test/utils/setup')],
+  require: [
+    require.resolve('@babel/register', { extensions: ['.js', '.ts', '.tsx'] }),
```
Comment (oliviertassinari) [inline .mocharc.js:5]: @eps1lon has done a few changes in https://github.com/mui-org/material-ui/pull/22179 that might help :)

File: .eslintrc.js [hunk 2]
```diff
@@ -220,7 +220,7 @@ module.exports = {
       },
     },
     {
-      files: ['*.spec.tsx', '*.spec.ts'],
+      files: ['*.spec.tsx', '*.spec.ts', "*.test.tsx", " *.test.ts"],
```
Comment (oliviertassinari) [inline .eslintrc.js:223]: It sounds great. We use "spec" for end-to-end test on TypeScript, only and "test" for the end-to-end functional test :)

File: test/utils/setup.js [hunk 3]
```diff
@@ -72,7 +77,7 @@ function throwOnUnexpectedConsoleMessages(methodName, expectedMatcher) {
   });
 }
 
-throwOnUnexpectedConsoleMessages('warn', 'toWarnDev');
-throwOnUnexpectedConsoleMessages('error', 'toErrorDev');
+// throwOnUnexpectedConsoleMessages('warn', 'toWarnDev');
```
Comment (oliviertassinari) [inline test/utils/setup.js:80]: @eps1lon This line was commented after Dmitriy mentioned that errors were important swallows. I have faced this problem too when adding the TypeScript tests on the material-ui-x repository. Now, it only happens when working on the transpilation, which is not frequent. So, one could argue that it doesn't need extra care. Not sure, at least, it's worth raising :).

**Comment by dmtrKovalenko** [comment]
I'd say that proper title is migrate tests to mocha. Because it is the main pain :)

**Comment by oliviertassinari** [comment]
Why is mocha a pain? I was under the assumption that the complexity is in testing library :)

File: test/utils/setup.js (hunk 3 above)
Comment (eps1lon) [reply test/utils/setup.js:80]: Repro would be nice. I don't understand what kind of errors would be "swallowed".

**Comment by eps1lon** [comment]
Mocha is not helpful if the underlying infra is "wrong" e.g. syntax errors in setup files let mocha exit with non-zero but show no error message whatsoever. So I understand the pain. I'd rather have a thousand errors with a message than a single one without any message.

File: test/utils/setup.js (hunk 3 above)
Comment (oliviertassinari) [reply test/utils/setup.js:80]: I can't figure out how to reproduce anymore. Maybe Dmitriy will

Comment (dmtrKovalenko) [reply test/utils/setup.js:80]: Any parsing/loading file issues. Originally met this by importing `ts` file from `.js`. (replaced one of the internal SVG icons with ts version).

But probably the easiest reproduction is running tests with merge conflicts: 

With commented lines:
image

Without: 
image

Comment (eps1lon) [reply test/utils/setup.js:80]: Awesome, thanks. Seems like mocha is using `console.error` internally? I think I have a better idea for that. Pretty sure we had this issue forever since we always stubbed console.error.

**Comment by eps1lon** [comment]
> eps1lon closed this in mui-org/material-ui#22196 1 hour ago

Sorry, need to remember that GitHub ignores repo boundaries.

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx [hunk 4]
```diff
@@ -0,0 +1,223 @@
+import * as React from 'react';
+import { expect } from 'chai';
+import { spy } from 'sinon';
+import TextField from '@material-ui/core/TextField';
+import { fireEvent, screen, act } from 'test/utils';
+import { StaticDatePicker, DesktopDatePicker, DesktopDatePickerProps } from './DatePicker';
+import { createPickerRender, adapterToUse } from '../internal/pickers/test-utils';
+import { CalendarSkeleton } from '../CalendarSkeleton';
```
Comment (oliviertassinari) [inline packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:8]: For public components, let's us the public import

```suggestion
import CalendarSkeleton from '@material-ui/lab/CalendarSkeleton';
```

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+      {...props}
+    />
+  );
+}
+
+describe('<DatePicker /> keyboard interactions', () => {
+  const render = createPickerRender({ strict: false });
+
+  context('input', () => {
+    function typeInPicker(value: string) {
```
Comment (oliviertassinari) [inline packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:32]: I think that we should drop this helper. We tend to duplicate (low abstractions) as much as possible between the tests. @eps1lon what do you think?

File: packages/pickers/lib/package.json [hunk 6]
```diff
@@ -35,10 +35,7 @@
     "@material-ui/core": "^5.0.0-alpha.5",
     "@material-ui/lab": "^5.0.0-alpha.5",
     "@types/react": "^16.8.6",
-    "date-fns": "^2.0.0",
-    "dayjs": "^1.8.17",
-    "luxon": "^1.21.3",
-    "moment": "^2.24.0",
+
```
Comment (oliviertassinari) [inline packages/pickers/lib/package.json:38]:
```suggestion
```

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    for (let i = 0; i < 3; i++) {
+      fireEvent.keyDown(document.body, { force: true, keyCode: 37, key: 'ArrowLeft' });
+    }
+
+    // leaves focus on the same date
+    expect(document.activeElement).to.have.attr('aria-label', 'Aug 13, 2020');
+  });
+
+  context('YearPicker keyboard navigation', () => {
+    beforeEach(() => {
```
Comment (oliviertassinari) [inline packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:176]: Would it work if we move the before each in the test itself (it)?

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+      expect(screen.getByRole('textbox')).to.have.value('10/11/2018');
+      expect(onChangeMock.callCount).to.equal(1);
+    });
+
+    it("Doesn't accept invalid date format", () => {
+      render(<TestKeyboardDatePicker />);
+
+      typeInPicker('01');
+      expect(screen.getByRole('textbox')).to.have.attr('aria-invalid', 'true');
```
Comment (oliviertassinari) [inline packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:57]: Please grep in the codebase to seek for what's most popular and use it, before introducing a new API. They are aliases: https://github.com/nathanboktae/chai-dom#attrname-value.

```suggestion
      expect(screen.getByRole('textbox')).to.have.attribute('aria-invalid', 'true');
```

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        target: { value },
+      });
+    }
+
+    it('Allows to change selected date from the input according to `format`', () => {
+      const onChangeMock = spy();
+      render(
+        <TestKeyboardDatePicker
+          onChange={onChangeMock}
+          inputFormat={adapterToUse.lib === 'moment' ? 'DD/MM/YYYY' : 'dd/MM/yyyy'}
```
Comment (oliviertassinari) [inline packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:43]: Should it be?

```suggestion
          inputFormat="dd/MM/yyyy"
```

File: packages/material-ui-lab/src/DatePicker/DatePickerLocalization.test.tsx [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        const input = screen.getByRole('textbox');
+        fireEvent.change(input, { target: { value: invalid } });
+
+        expect(input).to.have.attribute('aria-invalid', 'true');
+      });
+
+      // Need to run with ICU loaded https://moment.github.io/luxon/docs/manual/install.html#node
+      if (adapterToUse.lib === 'luxon') {
+        return;
+      }
```
Comment (oliviertassinari) [inline packages/material-ui-lab/src/DatePicker/DatePickerLocalization.test.tsx:116]: ?
```suggestion
```

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+          displayStaticWrapperAs="desktop"
+          value={new Date('2020-08-13')}
+          onChange={() => {}}
+          renderInput={(props) => <TextField placeholder="10/10/2018" {...props} />}
+        />,
+      );
+    });
+
+    it('Autofocus selected day on mount', () => {
+      expect(document.activeElement).to.have.attr('aria-label', 'Aug 13, 2020');
```
Comment (oliviertassinari) [inline packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:133]: @eps1lon Should we use `.toHaveAccessibleName()` in these cases?

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx (hunk 5 above)
Comment (eps1lon) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:32]: I'll try to not use any helpers that are exclusive to a test. Usually they encode expected behavior or implementation details of the component. Once a test breaks you constantly go back and forth between the test and the helper. This indirection makes debugging IMO needlessly hart. Remember, it's more important to have readable code than code that is easy to write because we tend to read way more than write.

It's why I'm not a fan of libraries like `user-event`. When your implementation starts deviating from the expected behavior in the middle of a helper method of an outside module you're having a rough time.

Helper methods that are local to tests also tend to grow with time. If you have to debug a single test but have filter out all the other stuff that only other tests need you're wasting a lot of mental energy (again readability > writability).

The itch to abstract away is understandable (it's basically what programming is about). But I've rarely seen this payoff in tests. You can always add these helpers initially while writing but when you want to merge inlining these will payoff longterm.

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx (hunk 4 above)
Comment (dmtrKovalenko) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:8]: It doesn't work for now because I didn't add public files :)

File: packages/pickers/lib/package.json (hunk 6 above)
Comment (dmtrKovalenko) [reply packages/pickers/lib/package.json:38]: This file will be deleted anyway, probably we need to revert all the changes here

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx (hunk 9 above)
Comment (dmtrKovalenko) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:43]: No, because we must support running tests with different adapters at least locally

File: packages/material-ui-lab/src/DatePicker/DatePickerLocalization.test.tsx (hunk 10 above)
Comment (dmtrKovalenko) [reply packages/material-ui-lab/src/DatePicker/DatePickerLocalization.test.tsx:116]: We must support diferrent libs

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx (hunk 5 above)
Comment (dmtrKovalenko) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:32]: > Usually they encode expected behavior or implementation details of the component. Once a test breaks you constantly go back and forth between the test and the helper. This indirection makes debugging IMO needlessly hart. Remember, it's more important to have readable code than code that is easy to write because we tend to read way more than write.

I strongly disagree with this point overall, but probably for such simple tests like in the core, it makes sense :) 
I would remove this kind of event because it really looks hasty.

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx (hunk 9 above)
Comment (oliviertassinari) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:43]: > adapters at least locally

IMHO, either it runs in the CI or not at all. I think that tests that only run locally should be deleted.

Comment (dmtrKovalenko) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:43]: I mean that running tests for all adapters not only for date-fns saved my ass from the library-specific issues a lot of times (well I mean more than 10 or 50 times tests were falling for specific adapters). 

I suppose that we discussed that this functionality will be implemented, but later.

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx (hunk 11 above)
Comment (eps1lon) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:133]: Wouldn't it make more sense assert what element is focused i.e. is it a button, textbox etc?

But generally whenever you assert on attributes responsible for naming you probably want `toHaveAccessibleName`. Though you could keep the attribute check for easier debugging in case this unit fails or documenting implementation choices.

If the goal is to have a fast test then I'd rather use simple data-testid and test naming in a dedicated test. Mixing these concerns is quite dangerous when you do refactoring. Say you change the keyboard focus behavior and move tests around or remove it. Now you also removed a test for naming and since we can't detect this kind of coverage loss you end up with untested behavior.

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx (hunk 9 above)
Comment (oliviertassinari) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:43]: If the overhead is small, why not. Otherwise, I think that the following would work great. 

1. We don't allow the component tests to run with different engines. We focus on date-fns exclusively, as we plan for the documentation.
2. We fully tests date adapter API, for each supported date library, in isolation, without integration. 
3. In the documentation, we have an example with each date library engine we support, as a smoke test for us and a way to quickly start or users.

Comment (dmtrKovalenko) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:43]: Not sure it will handle that corner cases that appeared, but I agree

File: packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx (hunk 11 above)
Comment (dmtrKovalenko) [reply packages/material-ui-lab/src/DatePicker/DatePickerKeyboard.test.tsx:133]: > Wouldn't it make more sense assert what element is focused i.e. is it a button, textbox etc?

I am doing this, didn't I? checking that `document.activeElement` has appropriate label name is pretty the same as looking for element and checking that it is equal to the `document.activeElement'

File: package.json [hunk 12]
```diff
@@ -39,7 +39,7 @@
     "test:regressions": "yarn test:regressions:build && rimraf test/regressions/screenshots/chrome/* && vrtest run --config test/vrtest.config.js --record",
     "test:regressions:build": "webpack --config test/regressions/webpack.config.js",
     "test:umd": "node packages/material-ui/test/umd/run.js",
-    "test:unit": "cross-env NODE_ENV=test mocha 'packages/**/*.test.js' 'docs/**/*.test.js' 'scripts/**/*.test.js' 'test/utils/**/*.test.js' --exclude '**/node_modules/**'",
+    "test:unit": "cross-env NODE_ENV=test mocha 'packages/**/*.test.js' 'packages/material-ui-lab/**/*.test.tsx' 'docs/**/*.test.js' 'scripts/**/*.test.js' 'test/utils/**/*.test.js' --exclude '**/node_modules/**'",
```
Comment (dmtrKovalenko) [inline package.json:42]: For now use additional grep for new test files. After migrated `pickers` folder will be removed I`ll change the grep to be `packages/**/*.test{ts,tsx,js}`
